src/test/java/org/broadinstitute/hellbender/engine/AssemblyRegionIteratorUnitTest.java

```diff
@@ -0,0 +1,115 @@
+
+public class AssemblyRegionIteratorUnitTest extends BaseTest {
+
+    @DataProvider
+    public Object[][] testCorrectRegionsHaveCorrectReadsAndSizeData() {
+        return new Object[][] {
+                { NA12878_20_21_WGS_bam, b37_reference_20_21, new SimpleInterval("20", 10000000, 10100000), 50, 300, 100 }
+        };
+    }
+
+    /*
+     * This test checks that over a 100,000 base interval, all assembly regions created by the AssemblyRegionIterator
+     * have the correct reads, that the reads are stored in the correct order, and that region padding and other
+     * settings are respected.
+     *
+     * We determine this by going by to the original BAM and doing a fresh query for each region, to ensure that
+     * the query results match the reads actually in the region.
+     */
+    @Test(dataProvider = "testCorrectRegionsHaveCorrectReadsAndSizeData")
+    public void testRegionsHaveCorrectReadsAndSize( final String reads, final String reference, final SimpleInterval shardInterval, final int minRegionSize, final int maxRegionSize, final int assemblyRegionPadding ) throws IOException {
+        try ( final ReadsDataSource readsSource = new ReadsDataSource(IOUtils.getPath(reads));
+              final ReferenceDataSource refSource = ReferenceDataSource.of(new File(reference)) ) {
+            final SAMSequenceDictionary readsDictionary = readsSource.getSequenceDictionary();
+            final LocalReadShard readShard = new LocalReadShard(shardInterval, shardInterval.expandWithinContig(assemblyRegionPadding, readsDictionary), readsSource);
+            final AssemblyRegionEvaluator evaluator = new HaplotypeCallerEngine(new HaplotypeCallerArgumentCollection(), false, false, readsSource.getHeader(), new CachingIndexedFastaSequenceFile(new File(b37_reference_20_21)));
+            final ReadCoordinateComparator readComparator = new ReadCoordinateComparator(readsSource.getHeader());
+
+            final List<ReadFilter> readFilters = new ArrayList<>(2);
+            readFilters.add(new WellformedReadFilter());
+            readFilters.add(new ReadFilterLibrary.MappedReadFilter());
+            final CountingReadFilter combinedReadFilter = CountingReadFilter.fromList(readFilters, readsSource.getHeader());
+            readShard.setReadFilter(combinedReadFilter);
+
+            final AssemblyRegionIterator iter = new AssemblyRegionIterator(readShard, readsSource.getHeader(), refSource, null, evaluator, minRegionSize, maxRegionSize, assemblyRegionPadding, 0.002, 50);
+
+            AssemblyRegion previousRegion = null;
+            while ( iter.hasNext() ) {
+                final AssemblyRegion region = iter.next();
+
+                Assert.assertTrue(region.getSpan().size() <= maxRegionSize, "region size " + region.getSpan().size() + " exceeds the configured maximum: " + maxRegionSize);
+
+                final int regionContigLength = readsDictionary.getSequence(region.getSpan().getContig()).getSequenceLength();
```

Comment:
Do you ever assert that there are no gaps between regions? I see that you assert their ordering and their overhangs, but what if they overlap/have gaps.

Reply:
Added